src/Main.purs

```diff
@@ -85,34 +117,70 @@ readMode = case _ of
 readFiles = lmap show <<< runExcept <<< traverse readString
 
-processFile :: Config -> String -> ExceptT String Aff Unit
+processFile :: Config -> String -> ExceptT String Aff (Tuple String (Array Module))
 processFile config fileName = do
   contents <- withExceptT Error.message
                 <<< ExceptT
                 <<< liftEffect
                 <<< try
                 $ Sync.readTextFile UTF8 fileName
-  modules <- except $ lmap (errorMessage fileName) (runParser contents wholeFile)
-  processModules config fileName modules
+  except $ lmap (errorMessage fileName) (map (Tuple fileName) (runParser contents wholeFile))
 
 processModules :: Config -> String -> Array Module -> ExceptT String Aff Unit
 processModules config fileName modules = do
-  output <- case config.mode of
-    Pretty -> pure $ PrintPrinter.prettyPrint modules
-    Purs -> pure $ Purescript.prettyPrint config.package modules
-    Scala -> pure $ Scala.prettyPrint config.package modules
-    Show -> pure $ show modules
+  case config.mode of
+    Pretty -> writeOutput config modules PrettyPrint.outputSpec
+    Purs -> writeOutput config modules (Purescript.outputSpec config.package)
+    Scala -> writeOutput config modules (Scala.outputSpec config.package)
+    Show -> Console.info $ show modules
     Test -> do
       b <- except $ lmap (errorMessage fileName) (roundTrip modules)
       if b
-        then pure "Round-trip passed"
+        then Console.info "Round-trip passed"
         else except $ Left "Round-trip failed"
-  Console.info output
+
+liftEffectSafely :: forall a. Effect a -> ExceptT String Aff a
+liftEffectSafely = ExceptT <<< liftEffect <<< map (lmap Error.message) <<< try
+
+outputPath :: Config -> Maybe (Array String)
+outputPath config =
+  config.outputDirectory <#> \o ->
+    let packagePath = String.replaceAll
+          (String.Pattern ".")
+          (String.Replacement "/")
+          config.package
+    in Array.filter
+        (\s -> String.length s > 0)
+        (String.split (String.Pattern "/") (o <> "/" <> packagePath))
+
+outputDirectories :: Array String -> Array String
+outputDirectories = scanl (\b a -> if b == "" then a else b <> "/" <> a) ""
+
+writeOutput :: Config -> Array Module -> OutputSpec -> ExceptT String Aff Unit
+writeOutput config modules outputSpec = liftEffectSafely do
+  for_ modules (\mod ->
+    outputPath config # maybe
+      (Console.info <<< scrubEolSpaces <<< outputSpec.render $ mod)
+      (writeOutput_ mod))
+  where
+    scrubEolSpaces :: String -> String
+    scrubEolSpaces i =
+      regex " +$" (Regex.Flags.multiline <> Regex.Flags.global) # either
+        (const i)
+        (\r -> Regex.replace r "" i)
+    writeOutput_ :: Module -> Array String -> Effect Unit
+    writeOutput_ mod dir =
+      Sync.writeTextFile
+        UTF8
+        (String.joinWith "/" filePath)
+        (scrubEolSpaces <<< outputSpec.render $ mod)
+      where
+        filePath = Array.snoc dir (outputSpec.fileName mod)
```

Comment:
I think it would be best if the build system can control the output filename instead of having it derived from the contents of the input file. Depending on `purs` performance, we might just have to live with that though.